pkg/healthcheck/healthcheck.go

```diff
@@ -382,38 +452,109 @@ func (hc *HealthChecker) allCategories() []category {
 					hintAnchor:  "l5d-existence-cr",
 					fatal:       true,
 					check: func(context.Context) error {
-						return hc.checkClusterRoles()
+						return hc.checkClusterRoles(true)
 					},
 				},
 				{
 					description: "control plane ClusterRoleBindings exist",
 					hintAnchor:  "l5d-existence-crb",
 					fatal:       true,
 					check: func(context.Context) error {
-						return hc.checkClusterRoleBindings()
+						return hc.checkClusterRoleBindings(true)
 					},
 				},
 				{
 					description: "control plane ServiceAccounts exist",
 					hintAnchor:  "l5d-existence-sa",
 					fatal:       true,
 					check: func(context.Context) error {
-						return hc.checkServiceAccounts()
+						return hc.checkServiceAccounts(true)
 					},
 				},
 				{
 					description: "control plane CustomResourceDefinitions exist",
 					hintAnchor:  "l5d-existence-crd",
 					fatal:       true,
 					check: func(context.Context) error {
-						return hc.checkCustomResourceDefinitions()
+						return hc.checkCustomResourceDefinitions(true)
+					},
+				},
+				{
+					description: "control plane MutatingWebhookConfigurations exist",
+					hintAnchor:  "l5d-existence-mwc",
+					fatal:       true,
+					check: func(context.Context) error {
+						return hc.checkMutatingWebhookConfigurations(true)
+					},
+				},
+				{
+					description: "control plane ValidatingWebhookConfigurations exist",
+					hintAnchor:  "l5d-existence-vwc",
+					fatal:       true,
+					check: func(context.Context) error {
+						return hc.checkValidatingWebhookConfigurations(true)
+					},
+				},
+				{
+					description: "control plane PodSecurityPolicies exist",
+					hintAnchor:  "l5d-existence-psp",
+					fatal:       true,
+					check: func(context.Context) error {
+						return hc.checkPodSecurityPolicies(true)
+					},
+				},
+			},
+		},
+		{
+			id: LinkerdConfigNotExistsChecks,
+			checkers: []checker{
+				{
+					description: "control plane namespace exists",
+					hintAnchor:  "l5d-control-plane-ns-exists",
+					fatal:       true,
+					check: func(context.Context) error {
+						if hc.kubeAPI == nil {
+							var err error
+							hc.kubeAPI, err = k8s.NewAPI(hc.KubeConfig, hc.KubeContext, requestTimeout)
+							if err != nil {
+								return err
+							}
+						}
+
+						return hc.CheckNamespace(hc.ControlPlaneNamespace, true)
+					},
+				},
+				{
+					description: "'linkerd-config' config map does not exist",
+					hintAnchor:  "l5d-linkerd-config-not-exists",
+					fatal:       true,
+					check: func(context.Context) error {
+						// to successfully install the control plane, the `linkerd-config`
+						// config map must not exist.
+						if err := hc.checkLinkerdConfigConfigMap(); !kerrors.IsNotFound(err) {
```

Comment:
since `checkLinkerdConfigConfigMap` is namespace-specific, can we remove the `l5d-control-plane-ns-exists` check above?

taking this a step further, since `LinkerdConfigNotExistsChecks` is not used by the `linkerd check` command (and thus does not require all the category/hintanchor/check()/etc machinery that health checkers provide, it may be more straightforward to simply implement the `linkerd-config` check in terms of the existing `fetchConfigs` function:

```diff
diff --git a/cli/cmd/install.go b/cli/cmd/install.go
index 701d63ee..49bd09e3 100644
--- a/cli/cmd/install.go
+++ b/cli/cmd/install.go
@@ -23,6 +23,7 @@ import (
 	log "github.com/sirupsen/logrus"
 	"github.com/spf13/cobra"
 	"github.com/spf13/pflag"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/util/validation"
 	"k8s.io/helm/pkg/chartutil"
 	"k8s.io/helm/pkg/proto/hapi/chart"
@@ -853,17 +854,14 @@ func globalResourcesExist() bool {
 }
 
 func linkerdConfigConfigMapExists() bool {
-	checks := []healthcheck.CategoryID{
-		healthcheck.LinkerdConfigNotExistsChecks,
+	kubeAPI, err := k8s.NewAPI(kubeconfigPath, kubeContext, 0)
+	if err != nil {
+		return false
 	}
 
-	hc := healthcheck.NewHealthChecker(checks, &healthcheck.Options{
-		ControlPlaneNamespace: controlPlaneNamespace,
-		KubeConfig:            kubeconfigPath,
-	})
+	_, err = fetchConfigs(kubeAPI)
 
-	notExists := hc.RunChecks(func(result *healthcheck.CheckResult) {})
-	return !notExists
+	return !kerrors.IsNotFound(err)
 }
 
 func (idopts *installIdentityOptions) validate() error {
diff --git a/pkg/healthcheck/healthcheck.go b/pkg/healthcheck/healthcheck.go
index bf8222d3..ec27354f 100644
--- a/pkg/healthcheck/healthcheck.go
+++ b/pkg/healthcheck/healthcheck.go
@@ -8,8 +8,6 @@ import (
 	"strings"
 	"time"
 
-	kerrors "k8s.io/apimachinery/pkg/api/errors"
-
 	"github.com/linkerd/linkerd2/controller/api/public"
 	spclient "github.com/linkerd/linkerd2/controller/gen/client/clientset/versioned"
 	healthcheckPb "github.com/linkerd/linkerd2/controller/gen/common/healthcheck"
@@ -64,12 +62,6 @@ const (
 	// to determine if a control plane is already installed.
 	LinkerdPreInstallGlobalResourcesChecks CategoryID = "pre-linkerd-global-resources"
 
-	// LinkerdConfigNotExistsChecks adds a series of checks to determine
-	// if the `linkerd-config` config map already exist. In order to install a
-	// control plane in a namespace, this config map must not exist in that
-	// namespace.
-	LinkerdConfigNotExistsChecks CategoryID = "linkerd-config-config-map-not-exists"
-
 	// LinkerdConfigChecks enabled by `linkerd check config`
 
 	// LinkerdConfigChecks adds a series of checks to validate that the Linkerd
@@ -505,41 +497,6 @@ func (hc *HealthChecker) allCategories() []category {
 				},
 			},
 		},
-		{
-			id: LinkerdConfigNotExistsChecks,
-			checkers: []checker{
-				{
-					description: "control plane namespace exists",
-					hintAnchor:  "l5d-control-plane-ns-exists",
-					fatal:       true,
-					check: func(context.Context) error {
-						if hc.kubeAPI == nil {
-							var err error
-							hc.kubeAPI, err = k8s.NewAPI(hc.KubeConfig, hc.KubeContext, requestTimeout)
-							if err != nil {
-								return err
-							}
-						}
-
-						return hc.CheckNamespace(hc.ControlPlaneNamespace, true)
-					},
-				},
-				{
-					description: "'linkerd-config' config map does not exist",
-					hintAnchor:  "l5d-linkerd-config-not-exists",
-					fatal:       true,
-					check: func(context.Context) error {
-						// to successfully install the control plane, the `linkerd-config`
-						// config map must not exist.
-						if err := hc.checkLinkerdConfigConfigMap(); !kerrors.IsNotFound(err) {
-							return fmt.Errorf("found existing 'linkerd-config' config map")
-						}
-
-						return nil
-					},
-				},
-			},
-		},
 		{
 			id: LinkerdControlPlaneExistenceChecks,
 			checkers: []checker{
```

Reply:
The reason why `l5d-control-plane-ns-exists` is needed is to guard against cases like:
```
linkerd install config | kubectl apply -f -
linkerd install control-plane -l linkerd-dev | kubectl apply -f -
```
Without the namespace check, we won't be able to distinguish between the `linkerd-config` config map isn't created yet, and the namespace just doesn't exist.

I think moving this logic the `fetchConfigs` makes sense to me.